src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,338 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+import util.Utility;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In additional to representing a state,
+ * it contains the logic which handles toggling/highlighting of labels and simplifies the retrieval of
+ * the resulting status for UI to display.
+ */
+public class LabelPickerState {
+    Set<String> initialLabels;
+    List<String> addedLabels;
+    List<String> removedLabels;
+    List<String> matchedLabels;
+    OptionalInt currentSuggestionIndex;
+
+    /*
+     * Constructors
+     */
+
+    public LabelPickerState(Set<String> initialLabels) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = new ArrayList<>();
+        this.removedLabels = new ArrayList<>();
+        this.matchedLabels = new ArrayList<>();
+        this.currentSuggestionIndex = OptionalInt.empty();
+    }
+
+    private LabelPickerState(Set<String> initialLabels,
+                             List<String> addedLabels,
+                             List<String> removedLabels,
+                             List<String> matchedLabels,
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    private LabelPickerState(Set<String> initialLabels,
+                             List<String> addedLabels,
+                             List<String> removedLabels,
+                             List<String> matchedLabels) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.currentSuggestionIndex = OptionalInt.empty();
+    }
+
+    /**
+     * Determines resulting state based on given initial state and user input
+     * @param initialState
+     * @param repoLabels
+     * @param userInput 
+     * @return new state that corresponds with the user input
+     */
+    public static LabelPickerState determineState(LabelPickerState initialState, Set<String> repoLabels,
```

Comment:
`repoLabels` need to be a parameter? Shouldn't it be an attribute of the state (i.e. a parameter to the constructor)?